[Advancement] Prototype/sepration from core

IamManchanda opened the pull request:
**Update**: This is Moved to https://github.com/zurb/foundation-sites/pull/10091

---------------------------------------------------------------------------------------------------------------

Implementation for #9736 

This adds helper classes for prototyping ( with responsive breakpoints ) but with a separate folder `scss/prototype` and is **disabled by default** (so, it doesn't heavy the framework) and is easily available with a simple 

```scss
@include foundation-prototype-classes;
```

To include only the specific prototype class ( eg, display ) you are looking for, you can use

```scss
@include foundation-prototype-display;
```

**Moreover you can also enable responsive breakpoints for prototype classes via sass variables.**
Responsive breakpoints are **disabled by default even if prototype mode is enabled**. You can turn it on through sass variables.

Special Note: For immutability purpose, Many Prototype classes use `!important` to ensure that these they aren't overriden by more specific selectors. As we all know that foundation framework conscientiously avoids using `!important` declarations. Also, note that i have only inserted `!important` on those specific CSS properties which i was fully sure that values of those properties shouldn't be overridden.

----------------------------------------------------------------------------------------------------------------------------

**Here is the task list**,
This prototype advance will gonna include:

**1. Current Classes** 
Added a depreciation notice as this will be added(moved) to prototypes in **v6.5** instead of **v6.4**. 
This is basically inline with the discussion i had with @kball in a **DM** on how to go about the current prototypes classes that are already there in the docs.

- [x] Float classes & clearfix ( #9723, Merged within this ) 
- [x] Visibility Classes
- [x] Text alignment classes

**2. New Classes**
Simply will be added in **v6.4** .... There are no second thoughts what so ever. 

- [x] Component Styling (Rounded, Bordered & Shadow)
- [x] Font Styling ( weight, italic, wide )
- [x] Title Bar Utility
- [x] Typescale
- [x] List Styling ( ordered, unordered )
- [x] Margin Helpers ( #9607 )
- [x] Padding Helpers ( #9607 )
- [x] Sizing Utilities 
- [x] Border box
- [x] Display classes ( #9719 )
- [x] Position Helpers
- [x] Overflow ( visibile, hidden, scroll )
- [x] Image Replacement, Text HIde ( #9718, Merged within this  )
- [x] Text transformation classes ( #9718, Merged within this  )
- [x] Text decoration classes
- [x] Text Truncate ( #9718, Merged within this  )
- [x] Text Wrapping ( #9718, Merged within this  )

**3. SASS only Mixins**
Besides classes, This also adds these new sass mixins (no generated css, only sass magic). Below is the list of all those newly added mixins which dont generate any css classes. 

- [x] Relational Mixins
- [x] Box Mixin 
- [x] Rotate Mixin 

**4. Documentation & Testing** 

- [x] Add a depreciation notice on current prototype specifc pages
- [x] Create a documentation page which includes all these prototypes classes and mixins ( currently new only in v6.4 ) 
- [x] Set up a Visual regression testing for all these new prototype classes.
- [ ] Add option to customizer

On Merge, this also closes https://github.com/protopen/protopen/issues/1 , infact the repo!

--------------------------------------------------------------------------------------------------------------------------------

**Please voice your opinions, advice, any other cool prototype suggestions and code reviews**,

IamManchanda committed:
Added text utilities(vanilla helper classes)
Additions
- Added text utility classes
- Added text transformation classes
Changes
- Imported and included these both above in _typography

IamManchanda committed:
End of line for _utilities

IamManchanda committed:
Added @group comments

IamManchanda committed:
Added Responsive variation for float classes.

IamManchanda committed:
Added a line space for better reading

IamManchanda committed:
Removed unnecessary parenthesis

IamManchanda committed:
Closes #9736 (Work in progress) Seprating prototype functionality
Changes (First commit) :
- Added a seprate `scss/prototype` folder in order to have clear sepration
- Added prototype variable as setting, false(disabled) by default!
- Created a `foundation-prototype-classes` mixin and included it within `_prototype.scss`
- Included `foundation-prototype-display` mixin within it with a `_display.scss` import
- Added Display classes with responsive breakpoint on: none, inline, inline-block, block, table, table-cell
- Added Display classes with responsive breakpoint on: flex, inline-flex (only enabled if flexbox is enabled)
- Included `foundation-prototype-classes` within docs

IamManchanda committed:
Merge branch 'feature/text-helpers' into prototype/sepration-from-core

IamManchanda committed:
Moved text utilities from typography to prototype

IamManchanda committed:
Created responsive breakpoints for text transformation

IamManchanda committed:
Correction in text tranformation

IamManchanda committed:
Responsive breakpoints for text-utilities

IamManchanda committed:
Merge branch 'feature/responsive-floats' into prototype/sepration-from-core

IamManchanda committed:
Moved `_float.scss` from components to prototype

IamManchanda committed:
Reordering within `_prototype.scss`.

IamManchanda commented:
Current CSS its adding on execute, 

1. Flexbox disabled & Prototype disabled ( default mode ) => no css 
2. Flexbox disabled & Prototype enabled => **[Gist](https://gist.github.com/IamManchanda/de6dfd6798d097f43d2403fc13e2c377)**
3. Flexbox enabled & Prototype enabled => **[Gist](https://gist.github.com/IamManchanda/211f96a273486202c20052cecba2d2fb)**

IamManchanda committed:
Moved spacing utility above sizing utility
No specific reasons, changed because this is one i am gonna work next.

IamManchanda committed:
Edit in @group comments

IllusionPerdu commented:
It's sould be good if for the "Float classes"  we can coose the breakpoint that we want the float. I think that in the majority of cases there is just need of the general "float class"
Same for the others who distinguish the different breakpoint.

IamManchanda committed:
Renamed mixin's name forthe sake of similarity

IllusionPerdu commented:
it's should be goo if we can have a map of the prototype we include ;)
I don't like when we include "all" or "none"

IamManchanda commented:
> it's should be goo if we can have a map of the prototype we include ;)
> I don't like when we include "all" or "none"

@kball @brettsmason  thoughts ?

IamManchanda committed:
End of line in `_display` and reordering within `_prototype`

IamManchanda committed:
Added text decoration mixin to prototype classes(including respective breakpoints
Includes
- text-overline
- text-underline
- text-line-through

IamManchanda committed:
Added font styling mixin to prototype classes(including respective breakpoints)
Includes
- font-normal
- font-bold
- font-italic

IamManchanda committed:
Correction for text-decoration

IamManchanda committed:
Added position helpers mixin to prototype classes (including respective breakpoints)
Includes
- position-static
- position-relative
- position-absolute
- position-fixed
- position-fixed-top
- position-fixed-bottom

IllusionPerdu commented:
@IamManchanda  You put the functionality of the float that was implemented in the prototype so those who used them are obliged to include the prototype and therefore have all of them;)
Since there is also the float for each "breakpoint" it does a lot of class in addition.

IamManchanda commented:
@IllusionPerdu Yeah i am thinking about that too of letting user's chose to use responsive breakpoints or not! Dont worry bro. It still a work in progress. Please bare with me!

natewiebe13 commented:
@IamManchanda a solution for this may be to add a global variable for creating responsive prototype classes (default to false), then for each file have a variable for that file that uses the global one by default. Similarly to how `$button-radius` works for example. Then by default you get helper classes that don't have responsive versions, but by changing a setting or two, you can enable them for all prototype classes, or turn them on individually.

IamManchanda commented:
Thanks @natewiebe13 but can you explain in brief by giving a coding example.... Thanks Harry!

IamManchanda commented:
@natewiebe13 Are you talking about something like `$global-prototype-responsive `

:arrow_down: 
```scss
$global-flexbox: true !global; 

@if $global-flexbox {
   @include flex-align($x: center, $y: top);
}
```

natewiebe13 commented:
@IamManchanda more like this:
```scss
// _global.scss
$global-prototype-responsive-classes: false !default;

// prototype/_float.scss
$prototype-float-responsive-classes: $global-prototype-responsive-classes !default;

@mixin foundation-prototype-float {
    @if ($prototype-float-responsive-classes) {
        @each $size in $breakpoint-classes {
            // Loop through breakpoints
        }
    }
    @else {
        // Non-responsive classes
    }
}
```

This way, if I wanted all of the prototype classes to be responsive, I could just set `$global-prototype-responsive-classes` to true, or if I only wanted the float prototype classes to be responsive, I could change `$prototype-float-responsive-classes` to true.

IamManchanda commented:
Very Good @natewiebe13 !

IllusionPerdu commented:
Here is a little how I see the things. (sorry for my bad english)

The config map in scss/prototype/_prototype.scss :

```
$proto: (
  float: (small,large),
  display: (medium,xxlarge)
);
....
@if map-has-key($proto, 'float') {
	@include foundation-prototype-float;
}
```

and in the file protype file float for example :

```
$floatbreak: map-get($proto, float);
@each $size in $breakpoint-classes {
	@if index($floatbreak,$size) {
		...
	}
}
```

IamManchanda commented:
sorry @IllusionPerdu but that's too much of over engineering for me ...... @natewiebe13 & others thought might be different though!

IamManchanda committed:
Added a global variable for responsive breakpoints, defaulted to false
`global-prototype-responsive-breakpoints`

IamManchanda committed:
Included floats into responsive breakpoints logic!

IamManchanda committed:
Included display into responsive breakpoints logic!

IamManchanda committed:
correction in `_display`

IamManchanda committed:
Included font styling into responsive breakpoints logic!

IamManchanda commented:
You know what .... i need to improve this logic ..... bit of dry coding here ....while implementing this .... just realised!

IamManchanda committed:
Improved logic within `_display`, `_float` and `_font-styling`
Now its more of a DRY Coding!

IamManchanda committed:
Included position into responsive breakpoints logic!

IamManchanda committed:
Correction in comments

IamManchanda committed:
Included text alignment into responsive breakpoints logic!

IamManchanda committed:
correction within text alignment

IamManchanda committed:
Included text decoration into responsive breakpoints logic!
Also removed an unnecessary line from text alignment

IamManchanda committed:
Correction in comments
Changed data type from number to boolean

IamManchanda committed:
Included text transformation and utilities into responsive breakpoints logic

IamManchanda commented:
@natewiebe13 Is this ohk now ..... is that what you were looking for ??  

@kball .... i hope you are liking where the code is heading ?

natewiebe13 commented:
@IamManchanda, yeah. That will help trim down on the amount of CSS generated unless needed. :+1:

IamManchanda commented:
@natewiebe13 But it didn't helped me to trim down my no of commits :stuck_out_tongue: lol

@DaSchTour How can we implement color utilities now, will you open a seprate PR for that after this is merged or within this only you will guide??

kball commented:
One thing I'd like to raise is that for the mixins & classes that are *currently included* in Foundation by default, it would be good to have them separately included with a default inclusion that prints a deprecation notice.

Of the items already addressed in this PR that would be the `foundation-float-classes`, but I think there are other places within the codebase that can and should be classified as prototype-specific.

IamManchanda commented:
Ohk @kball ..... have updated my task list above with addition helpers and will work on it!
{ Subheader, Lead Paragraph, Un-bulleted List, Statistics, Visibility Classes, Responsive Embed }

If you wanna add any other, please tell me 
Also, Things to consider => Thumbnails, callouts ???

-------------------------------------------------------------------------------------------------------------

Also, i would suggest to create a specific Prototype page
on similar lines to => http://foundation.zurb.com/sites/docs/kitchen-sink.html

brettsmason commented:
@IamManchanda I think a few of the ones you mentioned probably wouldn't fall under prototyping, but I could be wrong.

Responsive embed, subheader, lead paragraph, un-bulleted list and maybe statistics?

Others will have a view too though.

IamManchanda commented:
Ohk @brettsmason ..... I will not start work on them for the time being then ..... Let other have there opinions ...... No issue

IamManchanda committed:
Added list-style-type (ordered and unordered) classes to prototype mode
This Adds these below classes:
Unordered: `ul.list-none`, `ul.list-disc`, `ul.list-circle`, `ul.list-square`
Ordered: `ol.list-none`, `ol.list-decimal`, `ol.list-lower-alpha`, `ol.list-lower-latin`, `ol.list-lower-roman`, `ol.list-upper-alpha`, `ol.list-upper-latin`, `ol.list-upper-roman`

IamManchanda committed:
End of line inserted into `_list-style-type.scss`

IamManchanda commented on scss/prototype/_display.scss:
```diff
@@ -0,0 +1,50 @@
+$prototype-display-responsive-breakpoints: $global-prototype-responsive-breakpoints !default;
+
+@mixin foundation-prototype-display {
+  @each $display in (none, inline, inline-block, block, table, table-cell) {
```
@brettsmason need your advice here 
I am thinking of adding a map variable like `$display-map`
Here's what i am looking for, 

```scss
$display-map: (
  none, inline, inline-block, block, table, table-cell
)!default;

@each $display in $display-map {
  /* All my styling here */
}
```

IllusionPerdu replied:
yes !

IamManchanda replied:
Thanks @IllusionPerdu ..... but will wait for **brett or nate or kevin**'s answer for it .....
as i don't wanna do rework and all!

natewiebe13 replied:
@IamManchanda, I think that makes sense. This way it's more customizable. One thing to remember is to make sure the variables aren't too generic like `$display-map` as by the name alone, there's no indication that it's specifically for prototyping. Other than that :+1:

IamManchanda replied:
@natewiebe13 so how about `$display-all` or `$prototype-display-all` ?

natewiebe13 replied:
Maybe `$prototype-display-types`?

IamManchanda replied:
Hmmmnnn,,,,, @natewiebe13 @IllusionPerdu but,
Where where do i add these and where i leave??
In `display`, `list-style-type` , `text-decoration`, `text-transformation` it makes sense because there it should be customized.

But what about `float` and `text-alignment`  
- Float has (left, right)
- Text alignment has (left, right, center, justify)

Will it make sense to add in these two also, because you don't want user to customize that??

natewiebe13 replied:
@IamManchanda, I don't think there's much downside to having it customizable? The files are relatively small at the moment anyways.

IllusionPerdu replied:
For me, anything that increases the size of the css and that is not necessarily used should be optional.

IllusionPerdu replied:
for example if we don't use text-justify they are 215 octect who don't use if we use only the small, medium and large breakpoint: 
```
.text-justify {text-align: justify;}
@media print, screen and (min-width: 39em) {.medium-text-justify {text-align: justify;}}
@media print, screen and (min-width: 60em) {.large-text-justify {text-align: justify;}}
```

IamManchanda committed:
Moved visibility classes from components to visibility
Moved: `scss/components/_visibility.scss` -> `scss/prototype/_visibility.scss`

IamManchanda committed:
Added map variables within `_display`. Also improves logic
Variables(map) added:
- prototype-display-types : none, inline, inline-block, block, table, table-cell
- prototype-display-flex-types : flex, inline-flex

IamManchanda commented:
@natewiebe13 i am changing it to `$prototype-display-classes` instead of types
**Reason**: on `_list-style-type` thing like 
=> `$prototype-style-type-unordered-types` looking confusing so that's why changing

IamManchanda committed:
Changed types to classes within `_display`.
New names of renamed Variables(map):
- prototype-display-classes
- prototype-display-flex-classes

IllusionPerdu commented:
I think the maximun of optional should be great beacaus for exemple if we don't use the text-align:justify
there are 215o we don't use if we use only the defaut breakpoint... the file size is important...

```
.text-justify {text-align: justify;}
@media print, screen and (min-width: 39em) {.medium-text-justify {text-align: justify;}}
@media print, screen and (min-width: 60em) {.large-text-justify {text-align: justify;}}
```

IamManchanda commented:
@IllusionPerdu Since IE 10 ..... 

- A sheet may contain up to 65534 rules
- A document may use up to 4095 stylesheets
- @import nesting is limited to 4095 levels (due to the 4095 stylesheet limit)

IllusionPerdu commented:
**big file => more time to dowload file => penality from google/bind/... because low speed site :s** and this only for code who don't use :s

and the foundation6 is from ie9+ ;)

The rules for IE9 are:
* A sheet may contain up to 4095 selectors
* A sheet may @import up to 31 sheets
* @import nesting supports up to 4 levels deep

IamManchanda commented:
@IllusionPerdu 
- First of all, we are not adding much css !
- Second , Responsive breakpoints are disabled by default now even if prototype mode is enabled!
- Third to include only the specific prototype ( eg. float ) instead of all, you can simply use: 

```scss
@include foundation-prototype-float;
```
instead of 

```scss
@include foundation-prototype-classes;
```

IllusionPerdu commented:
yes and the webmaster can also write the class who he use... Especially when it's just to set a css property...

Let's add also ;)
prototype-color
```
@for $r from 0 through 255 {
  @for $g from 0 through 255 {
   @for $b from 0 through 255 {
      .color-#{$r}-#{$g}-#{$b} {
        color: rgb(#{$r}, #{$g}, #{$b});
      }
    }
  }
}
@for $r from 0 through 255 {
  @for $g from 0 through 255 {
   @for $b from 0 through 255 {
      .background-color-#{$r}-#{$g}-#{$b} {
        background-color: rgb(#{$r}, #{$g}, #{$b});
      }
    }
  }
}
```

IamManchanda commented:
@IllusionPerdu ..... Color utilities will be taken care by @DaSchTour 
and don't worry that will be not like your above example!

IamManchanda committed:
Added map variables within `_list-style-type`. Also improves logic
Variables(map) added:
- prototype-style-type-unordered-classes : none, disc, circle, square
- prototype-style-type-ordered-classes : none, decimal, lower-alpha, lower-latin, lower-roman, upper-alpha, upper-latin, upper-roman

IamManchanda committed:
Added map variables within `_float`. Also improves logic
Variables(map) added:
- prototype-float-classes : left, right

IamManchanda committed:
Added map variables within `_text-alignment`. Also improves logic
Variables(map) added:
- prototype-text-alignment-classes : left, right, center, justify

IamManchanda committed:
Added map variables within `_position`.
Variables(map) added:
- prototype-position-classes : static, relative, absolute, fixed

IamManchanda committed:
Added map variables within `_decoration`.
Variables(map) added:
- prototype-text-decoration-classes : overline, underline, line-through

IamManchanda committed:
Added map variables within `_transformation`.
Variables(map) added:
- prototype-text-transformation-classes : lowercase, uppercase, capitalize

IamManchanda commented:
@natewiebe13 @brettsmason  i am trying to code spacing utilities,
Here's below what i did but this is not working ? 
I dont get any error but code is not working 
Do i have to do something in sass functions ? ( `/scss/util/*` folder ) 

```scss
$prototype-spacer: 1rem;
$prototype-spacer-x: $prototype-spacer;
$prototype-spacer-y: $prototype-spacer;

$prototype-spacers: ();

@mixin foundation-prototype-spacing {
  @for $i from 0 through 3 {
    $prototype-spacers: map-merge($prototype-spacers, (
      $i: (
        x: $prototype-spacer-x * $i,
        y: $prototype-spacer-y * $i
      )
    )) 
  }

  @each $prop, $abbrev in (margin: m, padding: p) {
    @each $size, $lengths in $prototype-spacers {
      $length-x:   map-get($lengths, x);
      $length-y:   map-get($lengths, y);

      .#{$abbrev}-a-#{$size} { #{$prop}:        $length-y $length-x !important; } // a = All sides
      .#{$abbrev}-t-#{$size} { #{$prop}-top:    $length-y !important; }
      .#{$abbrev}-r-#{$size} { #{$prop}-right:  $length-x !important; }
      .#{$abbrev}-b-#{$size} { #{$prop}-bottom: $length-y !important; }
      .#{$abbrev}-l-#{$size} { #{$prop}-left:   $length-x !important; }

      // Axes
      .#{$abbrev}-x-#{$size} {
        #{$prop}-right:  $length-x !important;
        #{$prop}-left:   $length-x !important;
      }
      .#{$abbrev}-y-#{$size} {
        #{$prop}-top:    $length-y !important;
        #{$prop}-bottom: $length-y !important;
      }
    }
  }
}
```

Code ref : https://gist.github.com/lazabogdan/6878954076420348fb4f

IamManchanda committed:
Reordering: moving current prototype classes to top

IamManchanda committed:
More reordering for current prototype classes available

IamManchanda commented:
@kball ..... brett said,

> @IamManchanda I think a few of the ones you mentioned probably wouldn't fall under prototyping, but I could be wrong
> 
> Responsive embed, subheader, lead paragraph, un-bulleted list and maybe statistics?

So, just wanna know are they prototype specific classes or not in your view?

kball commented:
I don't really think typography falls under the headline of purely prototyping... and responsive embed definitely not. What do you mean by statistics?

IamManchanda commented:
http://foundation.zurb.com/sites/docs/typography-helpers.html#statistics @kball

IamManchanda committed:
Reloacated old code back for float, text alignment and visibility classes
This is specific commit to let users dont have jitters. The new code for these prototypes has been created but won't be added untill 6.5 and the user's will be provided a clear cut depreciation notice that this will be removed from v6.5.
Within this commit i have added the old code back and commented out the new code for float, text alignment and visibility classes

IamManchanda committed:
End of line!

IamManchanda committed:
Added overflow classes to prototypes
Added classes: overflow-visible, overflow-hidden, overflow-scroll
Main(X) axis: overflow-x-visible, overflow-x-hidden, overflow-x-scroll
Cross(Y) axis: overflow-y-visible, overflow-y-hidden, overflow-y-scroll

IamManchanda committed:
Added a customized variable for `text-overflow` within `.text-truncate`
Available values for `text-overflow` are: clip, ellipsis, or a string

IamManchanda committed:
Corrections/Reordering!

IamManchanda committed:
Adds responsive sizing utilities to prototypes
Classes added
- width-25, width-50, width-75, width-100
- height-25, height-50, height-75, height-100
- With a simple map variable, you can add more(or remove/change) percentage classes as per your needs
- Also added `max-width-100 and `max-height-100`

IamManchanda committed:
Added two new css: `foundation-flex-prototype` && `foundation-prototype`
Specially made for non sass users who just use the css files instead.

IamManchanda commented:
@kball **Important** : I have added two new css files to the source => https://github.com/zurb/foundation-sites/pull/9738/commits/b0aa0e9

These two i have specially made for prototyper's as believe it or not, 
there are more css prototyper's than scss prototyper's who just like to download and use css files instead so this makes sense i guess.

But will keep the final decision on you !

IamManchanda committed:
End of line to source css files

IamManchanda committed:
Renamed `sizes` to `prototype-sizes`

IamManchanda committed:
Correction in data types comment

IamManchanda committed:
Added much awaited Spacing utilities

IamManchanda commented on scss/prototype/_spacing.scss:
```diff
@@ -0,0 +1,75 @@
+$prototype-spacing-classes: (
+  margin: m, 
+  padding: p
+)!default;
```
To whom so ever it may concern,

Spacing utilities share a different class convention then any other classes within this PR or the foundation framework
Eg, => `.m-t-10` 
Reason: It gets to head pretty quickly and easier for prototyping. Shall i keep it like or change it. 

Thoughts @natewiebe13

IamManchanda committed:
End of line for `_spacing.scss`

IamManchanda committed:
Added depreciation notice to float, visibility and text alignment

IamManchanda committed:
Added copyright and @group comments in `spacing`.

IamManchanda committed:
Added rounded utility to prototypes!
Really helpfull for images, buttons and more.

IamManchanda committed:
End of line for rounded utility

IamManchanda committed:
Removed the keyword responsive as breakpoints is enough.

IamManchanda committed:
Created new doc file `prototyping.md`

IamManchanda committed:
Renamed `text-overflow` to `prototype-text-overflow`

IamManchanda committed:
Renaming of variables and correction in comments for docs.

IamManchanda commented:
@brettsmason @natewiebe13  

In Spacing utility, i think that i haven't done things correct way so, i need your help 

1. There got to be better way than this => [**scss/prototype/_spacing.scss** (_**Line 22-47**_)](https://github.com/IamManchanda/foundation-sites/blob/51482dc803e550d7656c783ccb56b43b62a7036b/scss/prototype/_spacing.scss#L22-L47)

2. In Spacing utilities, i have used a different class convention then any other classes within this PR or the foundation framework ( basically different than foundation convention ) Eg, => `.m-t-10`. 
**Reason**: It gets to head pretty quickly and is easier for prototyping. Shall i keep it like this or change it to `margin-top-10` ?

IamManchanda committed:
`prototype.md`: Added introductory and enabling part

IamManchanda committed:
Added `protyotype.html` link to homepage and subnav

IamManchanda committed:
Added text helpers to the prototype docs
Added prototypes
- Text Transformation
- Text Decoration
- Font Styling
- Image Replacement
- Text Truncate
- Text Nowrap

IamManchanda committed:
Grammar corrections

IamManchanda committed:
Reordering and correction in the docs.

IamManchanda committed:
More corrections.. Now using h2 tag instead of h3 for docs

IamManchanda committed:
Added List styling to the docs. Unordered and ordered.

IamManchanda committed:
Added codepen example to text decoration.

IamManchanda committed:
Important: Removed the code & current prototype classes
Will rely on version control `Important` is the keyword

IamManchanda committed:
Added `rounded` class to the docs.

IamManchanda committed:
Grammatical changes in the docs

IllusionPerdu commented on scss/prototype/_display.scss:
```diff
@@ -0,0 +1,62 @@
+@mixin foundation-prototype-display {
+  @each $display in $prototype-display-classes {
+    .display-#{$display} {
+      display: $display !important;
```
I think it's soul'd be better if you create a mixing "foundation-display" and use it. 
```
@mixin foundation-display($display) {
    display: $display !important;
}
```

It would allow to easily be passed to the prototyping by class to sementic version.

IamManchanda replied:
Ohk @IllusionPerdu ..... I get your idea..... I will look into it .... thanks !

Also can you help me in above comment :arrow_double_up:  ...... https://github.com/zurb/foundation-sites/pull/9738#issuecomment-277518057

IllusionPerdu replied:
Personally I prefer the notation ".m-t-50" but I think it is better to respect the foundation convention ;)

IamManchanda replied:
What about question 1 @IllusionPerdu

IllusionPerdu replied:
The problem is that we are not the decision makers on the project so, perhaps with the agreement of the members of the foundation it is possible to apply a different "light" convention for everything regarding the prototype.

IllusionPerdu commented:
After reflection I think that for each prototype there should be an individual mixing in order to be able to prototype with class and to be easily passed to a sementic form.

IamManchanda committed:
Removed all codepen examples for time being
Have thinked about it ... will add codepen example of the prototypes that require it but only after 6.4 is released with this in a seprate PR.

IamManchanda committed:
Reordering of links within the page and subnav

IamManchanda committed:
Renamed Rounded to Rounded Elements

IamManchanda committed:
Removed `ul.list-none` & `ol.list-none` for Un-bulleted Lists
Improved unbulleted lists

IamManchanda committed:
Added position and display classes to docs and some reordering.

IamManchanda committed:
Added overflow classes to docs.

IamManchanda committed:
Added sizing utility to the docs

IamManchanda committed:
Removed flex prototypes as we will take that up in seprate PR

IamManchanda committed:
Removed flex from prototypes

IamManchanda committed:
Added spacing utilities to protypes docs

IamManchanda committed:
Edited introductory part

IamManchanda committed:
Added text-wrap

IamManchanda committed:
Added font typescale!

IamManchanda committed:
End of line to typescale

IamManchanda committed:
Improved mixin logic: Semantic + Dry Coding

IamManchanda commented:
@IllusionPerdu .... is that what you wanted?? ..... my above commit :arrow_double_up: ! https://github.com/zurb/foundation-sites/pull/9738/commits/93692b1781895c72b44e3d1613a86fc1734a0c39

IamManchanda committed:
Added font wide

IamManchanda committed:
Rename sizing classes to sizing

IllusionPerdu commented:
Yes i think it's good :)

IamManchanda committed:
Spacing Utilities: Added negative margin to the code!

IamManchanda committed:
Moved font-wide to font styling and added negative margin to docs.

IamManchanda committed:
Some more Correction in the docs.

IamManchanda committed:
Improved spacing utilities logic.

IamManchanda committed:
Renamed n to count!

IamManchanda committed:
Improved docs for spacing utilities.

kball commented:
I can see the argument for specific prototyping utility classes to include `!important` for immutability purposes (similar to has been done with float classes; these are a similar case), but would like to hear from more css/scss experts like @brettsmason, @andycochran, and @ncoden.

IamManchanda commented:
Ohk @kball, i will then on what they say and as asked above,
Are you ohk with this => `assets/foundation-prototype.scss`
{{ The css file i would like to generate after the merge }}

IamManchanda committed:
Added prototype-spacing-multiplier variable

IamManchanda committed:
Added docs comments for `prototype-spacer-variable`

IamManchanda committed:
Removed negative margin classes, as not so much useful

IamManchanda committed:
Removed negative margin from docs.

IamManchanda commented:
@kball @brettsmason ..... Have completed the scss and docs part ..... please review it for any changes. 

Note: Testing part to be completed soon

IamManchanda committed:
Visual Regression: Added Rounded elements to testing
Also, created a template specific to `prototype.css` for re-use.

IamManchanda committed:
Visual Regression: Added Font Styling to testing
Includes:
- Font Wide
- Font Normal
- Font Bold
- Font Italic

IamManchanda committed:
Visual Regression: Added font typescale to testing

IamManchanda committed:
Visual Regression: Added ordered and unordered lists to testing
Also removed `margin-left: 0` from scss as its was a mistake from my end.

IamManchanda committed:
Visual Regression: Added Sizing Utilities to testing
Inludes
- Width (25, 50, 75, 100)
- Height (25, 50, 75, 100)
- Max Width 100
- Max Height 100

IamManchanda committed:
Visual Regression: Added text wrap and no wrap to testing

IamManchanda committed:
Visual Regression: Added Text Truncate to testing

IamManchanda committed:
Visual Regression: Added Text Transformation to testing

IamManchanda committed:
Visual Regression: Added text decoration to testing

IamManchanda committed:
Visual Regression: Added image replacement (text-hide) to testing
Also updated docs for the same.

IamManchanda committed:
Visual Regression: Added display classes to testing
Also removed display-table and display-table-cell as default from docs and scss

IamManchanda committed:
Coorection in docs comments

IamManchanda committed:
Visual Regression: Add overflow to testing

TonyCodeGeek commented:
Woww
Me using foundation since v3
but this exctly wht was mising.

IamManchanda commented:
@TonyCodeGeek Thanks .... appreciated ..... !!!
Kev @kball ..... SCSS, Docs, Testing .... All done :muscle:

IamManchanda commented:
> I can see the argument for specific prototyping utility classes to include !important for immutability purposes (similar to has been done with float classes; these are a similar case), but would like to hear from more css/scss experts like @brettsmason, @andycochran, and @ncoden.

Ping @brettsmason awaiting answer !

brettsmason commented:
@IamManchanda sorry not ignoring you, just don't have much free time at the moment but I'll take a look as soon as I can :)

IamManchanda committed:
Added cards to rounded elements example
Within
- Foundation Docs
- Visual Regression Testing

IamManchanda committed:
Docs correction

IamManchanda committed:
Added bordered class to prototype utility!
Also
- Added bordered class to docs and visual regression testing.
- Added tables examples to docs and visual regression testing.
- Renamed Rounded elements to rounded & bordered

IamManchanda committed:
Add parameters to bordered mixin
- `width[prototype-border-width]` Width of the border
- `type[prototype-border-type]` Type of the border
- `color[prototype-border-color]` Color of the border

IamManchanda committed:
Added parameters to bordered mixins
- radius [prototype-border-radius] Border radius

IamManchanda committed:
Added paramaters/mixins docs to display classes
- display [] Display classes

IamManchanda committed:
Added parameters/docs mixins to list styling
- style-type-unordered [] Style type for unordered Lists
- style-type-ordered [] Style type for ordered Lists

IamManchanda committed:
Added Parameters/mixin docs to overflow classes
- overflow [] Overflow classes

IamManchanda committed:
Added paramters/mixin docs to position classes
- position [] Position classes
- z-index [prototype-position-z-index] z-index

IamManchanda committed:
Added mixin docs for `rounded-circle`
Also did some more grammatical docs correction

IamManchanda committed:
Added parameters/mixin docs to font styling
- letter-spacing [prototype-wide-letter-spacing] Wide letter spacing for the font
- weight [prototype-font-normal] Weight of the font (normal)
- weight [prototype-font-bold] Weight of the font (bold)

IamManchanda committed:
Added mixin docs to sizing utility
Also removed extra line space from spacing utility

IamManchanda committed:
Added parameters/mixin docs to text decoration and text transformation
- transformation [] Text Transformation
- decoration [] Text Decoration

IamManchanda committed:
Renamed some variables(map) names as per current code

IamManchanda committed:
Added parameter/mixin docs to text utilities
- overflow [prototype-text-overflow] Text Truncate

IamManchanda committed:
Improving docs for enabling prototype mode

IamManchanda commented:
@brettsmason ..... it's a no issue.... **you can take your time untill you're free** ... its really a no issueee 
But what i am looking for **at the moment** now is only on the clarity of `!important` .... That's it buddy!
#Please .... 

Reference => https://csswizardry.com/2016/05/the-importance-of-important/

brettsmason commented:
@IamManchanda I think in this case using `!important` is absolutely fine as long as its used sensibly, which you have 😄

IamManchanda committed:
Adding `!important` to **sensible** areas.

IamManchanda commented:
@brettsmason .... thanks just did that .... please note that i have only added `!important` where i am 100% sure that i dont want the user to edit the values.
For example,

```scss
.position-fixed-top {
  position: fixed !important;
  top: 0 !important;
  right: 0 !important;
  left: 0 !important;
  z-index: 975; 
}
```

Here in first 4 properties, i am 100% sure that i dont want user to change its values but i am happy to let them change `z-index`!! 

Hope you are happy with current implementation of `!important` ( above commit :arrow_double_up: ) .... if in case you are not with specific ones, please tell me, i will be happy to remove.

IamManchanda committed:
Documented a callout for `!important` within docs.

IamManchanda committed:
Added max-width-100 to text truncate

IamManchanda committed:
Hyperlinking the depreciation notice

IamManchanda committed:
Added relational mixin to prototyping utilities
Basically easy scss mixins to work upon nth selectors
This adds(or generates) no css whatsoever only scss mixins

IamManchanda commented:
@brettsmason ..... **Update**: My just Above commits adds Relational mixins to prototyping utilities. Please note that this doesn't adds any CSS what so ever, just simple scss mixins. 
Is it the right move or shall i revert ?

IamManchanda committed:
Added a box mixin to easily create a square, rectangle or a circle
Same like relational mixins, this also adds no generated css what so ever, just simple scss mixin.

IamManchanda committed:
Correction in sass mixins docs

IamManchanda committed:
Reordering, mixins first and then prototype classes

IamManchanda committed:
Added border box utility
Also added the docs for the same

IamManchanda committed:
End of line for border-box

IamManchanda committed:
Improved Position Mixin, better reusability!

IamManchanda committed:
Added margin and padding mixin for reusability!
No generated css whatsoever, only sass magic

IamManchanda committed:
Modified spacing utilities ( both classes and mixins )
Also updated docs

IamManchanda commented:
@DaSchTour, As per our discussion i have modified the [code](https://github.com/IamManchanda/foundation-sites/blob/a8adc4b74fb4dc10136f21908de9401b9ac7ffa9/scss/prototype/_spacing.scss#L36-L84) 

Also have updated the same thing in spacing classes and also have decreased the `spacers-count` from 5 to 3 as we are covering a lot now in the mixin!

Is this what you wanted ???

DaSchTour commented:
@IamManchanda that's what I suggested. This may than also be helpful as a global mixing to define spacing depending on given variables.

IamManchanda committed:
Spacing: Integrated margin and padding mixin to spacing classes
Also:
- Docs Updation
- Convention changed to foundation convention aka `padding-top-1` instead of `p-t-1`

IamManchanda commented:
Thanks @DaSchTour .... i have taken a step further and have integrated these mixins in my spacing classes and removed that old stupidity that i was having earlier. Here is the updated code => [_spacing.scss](https://github.com/IamManchanda/foundation-sites/blob/127d1e7a23f1d04ad08f8738a35555505c4cf988/scss/prototype/_spacing.scss#L63-L122)
Also updated my code to foundation convention!

I hope you will like it Daniel Schuba ??

IamManchanda committed:
Removed an extra line space!

IamManchanda committed:
Doc update: Addition in Sass includes!

IamManchanda committed:
Docs Update: Changed the text of `h4` tags as was creating same #link

IamManchanda committed:
Removed `all` keyword!

IamManchanda committed:
Optional responsive breakpoints for these spacing classes

IamManchanda committed:
Optional Responsive breakpoints for sizing classes!

IamManchanda committed:
Improved position mixin, added `global-position` variable & docs updated

IamManchanda committed:
Changed tagline

IamManchanda committed:
Fix minor indentation issues!

IamManchanda committed:
Improved helper box mixin and added it in docs

IamManchanda committed:
Rounded Utility: Removed unneccessary css !
Better to keep it simple for now even for sass mixins!

IamManchanda committed:
Docs update: Added Relational mixins.

IamManchanda committed:
Updated test pages with new margin and padding classes!

IamManchanda committed:
Renamed(and recoded) font typescale to typescale
`.font-h1` => `.h1` and so on like other frameworks!

IamManchanda committed:
Added a warning callout notice!
Prototyping classes like these below should only be used for prototyping tasks. Also if you are using **Sass**, we encourage you to use **[Prototyping Mixins](#sass-mixins)** instead.

IamManchanda committed:
Added component styling
Changes
- Moved `rounded` and `bordered` to component styling
- Created and added `shadow` to component styling

IamManchanda committed:
Added a Sass tip and a codepen!

IamManchanda committed:
Added a Heading Bar Utility
Added classes:
- `bar-left`
- `bar-center`
- `bar-right`

IamManchanda committed:
Docs edits

IamManchanda committed:
Renamed Heading bar utility to Title bar utility

IamManchanda committed:
Docs: Moved Image replacement(text-hide) to text helpers

IamManchanda committed:
Docs: Added example for text decoration!

IamManchanda committed:
Docs: Moved text helpers and Title bar Utility above

IamManchanda committed:
Title Bar Utility: Renamed the whole codebase to `titlebar` from `bar`

IamManchanda committed:
Docs: Renaming of the utility titles

IamManchanda committed:
Docs: Updated full list includes of prototype utilities.

IamManchanda committed:
Sass Mixin Helpers - Added a Rotate Mixin
No generated css what so ever.

IamManchanda committed:
Display Classes: Added `display-table`

IamManchanda committed:
Added foundation copyright & @group to box and rotate mixin.

IamManchanda committed:
Renamed (and recoded) titlebar to separator
As Titlebar conflicts with Top Bar => Title Bar

kball commented on scss/prototype/_border-box.scss:
```diff
@@ -0,0 +1,35 @@
+
+/// Border box utility
+@mixin border-box {
+	box-sizing: border-box !important;
```
whitespace off throughout this file - please adjust to 2 spaces per indent

kball commented on scss/prototype/_box.scss:
```diff
@@ -0,0 +1,23 @@
+@mixin box(
+	$width, 
+	$height: $width, 
+	$circle: false
```
please adjust whitespace to 2 spaces per indent

kball commented on scss/_global.scss:
```diff
@@ -74,6 +74,10 @@ $global-margin: 1rem !default;
 
+/// Global value used for positioning on components.
+/// @type Number
+$global-postion: 1rem !default;
```
typo - should be position

IamManchanda replied:
Accident ... fixed now!

kball commented:
Taking a look through this again... there's a ton of great prototyping helpers; I think we need a little more polish to get this across the line. I noted 1 typo, and also a lot of whitespace variability that needs to be fixed.

I'd also like to get another pass on the docs page copy... @rafibomb can you help with this?

IamManchanda committed:
Fixed indentation - Box Mixin!

IamManchanda committed:
Indetation Fix - Border Box Utilities!

IamManchanda committed:
Position Helpers - Fixed typo!

IamManchanda commented:
@kball The requested changes is done, now i have my 3 cents 

1. What to do with thumbnails ?? ( Same like float, visibility, text alignment or Keeping it in core)
2. I want to change the naming convention of margin and padding helper classes ( eg. m-x-0 instead of margin-horizontal-0 ) ... yes its different from foundation convention but quite useful with code brevity! ... What you say ?
3. Although a mixin, but i would like to show examples for every single mixin of Relational mixins but if i do that page will get longer .... moreover these relational mixin ( as no css involved ) doesn't fall purely under the category of Prototyping .... so why not have a separate section for it. Thoughts ? @rafibomb @kball 

Attaching an image of @include's of Relational Mixins for reference 
![relational-mixins](https://cloud.githubusercontent.com/assets/4970624/25195803/606c54ce-255c-11e7-805c-be332bcd94d2.png)

kball commented:
@rafibomb I know you were taking a look at this, did you have feedback for @IamManchanda?

IamManchanda commented:
@kball @rafibomb Moving ahead i think flexbox utilities should also be handled the same way also.
Same like prototyping classes, flexbox helper classes are also really only one step away from **inline styles**.

What we can do 
- [ ] Same like current classes (for eg text alignment). We can depreciate [Flexbox helper classes](http://foundation.zurb.com/sites/docs/flexbox.html#helper-classes) and move it later in 6.5
- [ ] Simply move [Vanilla helper classes](http://foundation.zurb.com/sites/docs/flexbox.html#vanilla-flexbox-helper-classes). I dont think any depreciation notice is needed here. Same like prototypes, we can add responsive breakpoints to a variable(`false` by default). Currently it has responsive breakpoints enabled by default which kinda sucks. 

----------------------------------------------------------------------------------------------------------------------------

BTW after the bold webpack move... Do we really need a depreciation notice ??
**I dont think so.**

Just give them option of choosing b/w these two files 
- You are a **designer/frontend dev**, choose `foundation.css/scss` and do the coding yourself. Use BEM/ITCSS/OOCSS or whatever. Make your own code, just take foundation as a building layer, these prototyping classes are just not for you! But yes if you are using Sass, these protototyping mixins can help you a lot!
- You are a **backend developer**, dont know designing... Use `foundation-prototype.css/scss`. Also checkout building blocks. We made that just for you!

Both community will be happy!

IamManchanda committed:
Try to resolve conflict - docs/pages/index.md

IamManchanda committed:
Resolve conflict - docs/partials/component-list.html

IamManchanda committed:
Closes #10045 - Add border none reset utility!

IamManchanda committed:
Correct grammar!

rafibomb commented on scss/prototype/_display.scss:
```diff
@@ -0,0 +1,46 @@
+/// Map containing all the `display` classes
+/// @type Map
+$prototype-display: (
+  none, inline, inline-block, block, table
```
do we need display none? have the `.hide` and visibility classes already.

rafibomb replied:
feels weird to leave `display: flex` out of this too. But we have a `.flex-container` class

IamManchanda replied:
I had it flex utilities before!
From my memory, @kball told me to remove flexbox things and keep it in a separate future PR as we were planning to make flexbox first in 6.4!

IamManchanda replied:
Hmmmnn.... You are right! 
Removing it!

IamManchanda committed:
Remove display none as we have hide!

IamManchanda committed:
Update docs on removing `display-none`

IamManchanda commented:
@rafibomb updated 

![display-new](https://cloud.githubusercontent.com/assets/4970624/26321163/64e456cc-3f43-11e7-8953-1957314315df.png)

rafibomb commented:
@kball This option will need to be added to the customizer.

IamManchanda commented:
Yes ofcourse and a file `foundation-prototype.css/scss` would be also good!

IamManchanda commented:
Closing this as moved to #10091! 
Cheers .... oh i am so happy!

peku33 commented:
@IamManchanda I really like the update, however I'm really unsure about moving `visibility classes` to prototyping.

In my opinion this feature is used in almost every scenario, even on the most basic websites.

Including anything from prototyping into final product turns on the red light in terms of stability or performance. 

I think it's enough for them to be moved into separate `foundation-visibility-classes` module, but making it a prototype feature? Are there any strong reasons for that?

IamManchanda commented:
Hey @peku33 ... First of all thankyou... Really means a lot to me. I am humbled!

The decision i made on this while back was inspiring from this comment by @tdhartwick [here](https://github.com/zurb/foundation-sites/issues/9607#issuecomment-270775555) 

But yes this will be no bad if we revisit this depreciation thing 

1. [Text Alignment](http://foundation.zurb.com/sites/docs/typography-helpers.html#text-alignment) : For this i dont think we need to revisit. 
2. [Visibility Classes](http://foundation.zurb.com/sites/docs/visibility.html) : We surely need to revisit on this.
3. [Float classes](http://foundation.zurb.com/sites/docs/float-classes.html) : Float and clearfix classes use become zero with flexbox based system... so we may need to remove this for XY Grid and Legacy Flex and just support this for Legacy Float Grid. Poke @kball 

That being said... There is no such that you need to enable or disable all prototypes classes. 
See http://foundation.zurb.com/sites/docs/prototyping-utilities.html#enabling-prototype-mode
For example you just want `separator` utility , you can just do 

```
@include foundation-prototype-separator;
```

peku33 commented:
The general idea about separating the core and prototyping functionality is great. Somewhere in February I started looking for new RWD solution. The main idea was to pick something lighter than bootstrap. All I needed was grid, nothing more. I gave foundation a try, because I was able to select only modules I needed without having to load all this heavy 'buttons / rotators / forms' stuff.

Application core was already made so I needed no prototyping functionality, just had to rewrite the frontend. The design was also ready, so all the components were already styled.

In the end foundation did everything I wanted it to do, but I used only:
`foundation-global-styles`
`foundation-grid` (I'm going to switch to the latest XY stuff in a moment)
`foundation-top-bar`
`foundation-visibility-classes`

I think global styles + grid + visibility classes are the real CORE, as they give no default style (look), just the skeleton, the real frontend base.

All the other stuff: buttons, forms, accordition, badges, etc are useless for me, because they are more _styles_ than framework. In my opinion most of projects that aim to use foundation are going to either use just the grid and provide their own look OR use everything you provide.

It would be great to separate non-visual classes as the core and visual components as prototyping.

I somehow expect that if use foundation in my project in which all styles and visual effects are provided, I have to use nothing from prototyping.
